backend/src/create.rs

```diff
@@ -2,118 +2,95 @@ use std::collections::{BTreeMap, BTreeSet, HashMap};
-                        node_ids.into_iter().map(|n| node_mapping[&n]).collect(),
-                    ));
-                }
-            }
+    }
+
+    fn way(
+        &mut self,
+        _: WayID,
+        node_ids: &Vec<NodeID>,
+        node_mapping: &HashMap<NodeID, Coord>,
+        tags: &Tags,
+    ) {
+        if node_ids.len() < 2 {
+            return;
+        }
+        if tags.has("railway") && (!tags.has("layer") || tags.is("layer", "0")) {
+            self.railways.push(LineString(
+                node_ids.into_iter().map(|n| node_mapping[&n]).collect(),
+            ));
+        } else if tags.is_any("natural", vec!["water", "coastline"]) || tags.is("waterway", "dock")
+        {
+            // If the entire area is inside the study area, the LineString will be closed. If
+            // it intersects the study area, then it might not be.
+            self.waterways.push(LineString(
+                node_ids.into_iter().map(|n| node_mapping[&n]).collect(),
+            ));
         }
-        Element::Relation { tags, members, .. } => {
-            let tags: Tags = tags.into();
-            if tags.is("type", "route") && tags.is("route", "bus") {
-                if let Some(name) = tags.get("name") {
-                    for (role, member) in members {
-                        if let osm_reader::OsmID::Way(w) = member {
-                            if role.is_empty() {
-                                bus_routes_on_roads
-                                    .entry(w)
-                                    .or_insert_with(Vec::new)
-                                    .push(name.to_string());
-                            }
+    }
+
+    fn relation(&mut self, _: RelationID, members: &Vec<(String, OsmID)>, tags: &Tags) {
+        if tags.is("type", "route") && tags.is("route", "bus") {
+            if let Some(name) = tags.get("name") {
+                for (role, member) in members {
+                    if let OsmID::Way(w) = member {
+                        if role.is_empty() {
+                            self.bus_routes_on_roads
+                                .entry(*w)
+                                .or_insert_with(Vec::new)
+                                .push(name.to_string());
                         }
                     }
                 }
             }
         }
-        Element::Bounds { .. } => {}
-    })?;
-
-    // There'll be many barrier nodes on non-driveable paths we don't consider roads. Filter for
-    // just those on things we consider roads.
-    let mut barrier_pts = Vec::new();
-    for way in &highways {
-        for node in &way.node_ids {
-            if all_barriers.contains(node) {
-                barrier_pts.push(node_mapping[node]);
-            }
-        }
     }
+}
+
+pub fn create_from_osm(
+    input_bytes: &[u8],
+    boundary_wgs84: Polygon,
+    study_area_name: Option<String>,
+) -> Result<MapModel> {
+    let mut osm = Osm::default();
```

Comment:
This function becomes more focused, now that extracting stuff from OSM is teased out